lib/SILOptimizer/Utils/OptimizerStatsUtils.cpp

```diff
@@ -0,0 +1,970 @@
+///       Symbol is e.g. the name of a function.
+///       StageName and TransformName are the names of the current optimizer
+///       pipeline stage and current transform.
+///       Duration is the duration of the transformation.
+//===----------------------------------------------------------------------===//
+
+#include <iostream>
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/Process.h"
+#include "swift/SIL/SILValue.h"
+#include "swift/SIL/SILVisitor.h"
+#include "swift/SILOptimizer/Analysis/Analysis.h"
+#include "swift/SILOptimizer/PassManager/PassManager.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/OptimizerStatsUtils.h"
+using namespace swift;
+
+namespace {
+
+/// The total number of different kinds of SILInstructions.
+constexpr int SILInstructionsNum = int(ValueKind::Last_SILInstruction) + 1;
+
+/// A set of counters, one per SILInstruction kind.
+typedef SmallVector<int, SILInstructionsNum> InstructionCounts;
+
+/// Map a SILInstruction name to its SILInstructionKind.
+SILInstructionKind getSILInstructionKind(StringRef InstName) {
+#define INST(Id, Parent, TextualName, MemoryBehavior, ReleasingBehavior)       \
+  if (InstName == #TextualName)                                                \
+    return SILInstructionKind::Id;
+
+#include "swift/SIL/SILNodes.def"
+
+#ifdef NDEBUG
+  std::cerr
+      << "Unknown insruction name used by the -sil-stats-only-instructions\n";
+  abort();
+#endif
+  llvm_unreachable(
+      "Unknown insruction name used by the -sil-stats-only-instructions");
+}
+
+/// Map SILInstructionKind to a corresponding SILInstruction name.
+const char *getSILInstructionName(SILInstructionKind Kind) {
+  switch (Kind) {
+#define INST(Id, Parent, TextualName, MemoryBehavior, ReleasingBehavior)       \
+  case SILInstructionKind::Id:                                                 \
+    return #TextualName;                                                       \
+    break;
+#include "swift/SIL/SILNodes.def"
+  }
+}
+
+/// A helper type to parse a comma separated list of SIL instruction names
+/// provided as argument of the -sil-stats-only-instructions options.
+class StatsOnlyInstructionsOpt {
+  /// If ComputeInstCounts[i] is non-zero, this kind of SILInstruction should be
+  /// tracked.
+  InstructionCounts ComputeInstCounts;
+  /// The number of different kinds of SILInstructions to be tracked.
+  int InstCountsNum = 0;
+
+public:
+  StatsOnlyInstructionsOpt() : ComputeInstCounts(SILInstructionsNum) {}
+
+  void operator=(const std::string &Val) {
+    if (Val.empty())
+      return;
+    if (Val == "all") {
+      for (auto &Inst : ComputeInstCounts) {
+        Inst = 1;
+      }
+      InstCountsNum = ComputeInstCounts.size();
+      return;
+    }
+    SmallVector<StringRef, 8> statsInstNames;
+    StringRef(Val).split(statsInstNames, ',', -1, false);
```

Comment:
Can't you use the incremental form of split:

http://llvm.org/doxygen/classllvm_1_1StringRef.html#a74783910a317456a47477f21dc8a73de

Then you don't need to use a small vector.

Reply:
Yes. But it is not worth the effort ;-) We are parsing command-line options here. The performance of this operation does not matter.

Reply:
Ok